CARDS-1983: Add related form link

acrowthe opened the pull request:
Add an optional `cards:links` child node to forms. This is displayed in the form header if present, either as breadcrumb links or a link with label if an optional label is present.

By default, a link is created for Heracles Pause/Resume forms. When a resume form is created, it will be linked to the most recent pause form and that pause form will link back to the new resume form. Adding additional pause/resume forms should result in more linked pairs of forms.

The label option can be tested using Composum by adding a 'label' property to an existing link node

![image](https://user-images.githubusercontent.com/20056751/235522968-6beaf491-7bdf-456b-a8b1-0c826739a3de.png)
![image](https://user-images.githubusercontent.com/20056751/235528182-9d1a3ef2-492e-4721-9eae-10ddaf7ab905.png)

To verify that links will not be exported, create a patient with pause/resume forms and check the json output that will be exported:
`http://localhost:8080/Subjects/<SubjectNodeName>.data.deep.bare.-labels.-identify.relativeDates.nolinks.json`

acrowthe committed:
CARDS-1983: Add related forms link

acrowthe committed:
CARDS-1983: Add related form link

Fix pause resume editor service user permission

acrowthe committed:
CARDS-1983: Add form reference between pause and resume forms
- Fix link from pause to resume form using a new listener

marta- commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/PauseResumeFormEditor.java:
```diff
@@ -54,6 +54,8 @@ public class PauseResumeFormEditor extends DefaultEditor
 {
     private static final Logger LOGGER = LoggerFactory.getLogger(PauseResumeFormEditor.class);
 
+    private static final String CREATED_STRING = "jcr:created";
```
```suggestion
    private static final String CREATED_PROP = "jcr:created";
```

marta- commented on heracles-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/Pause-Resume Status.xml:
```diff
@@ -50,11 +50,11 @@
-                        <value>formatted</value>
-                        <type>String</type>
-                </property>
+			<property>
+				<name>displayMode</name>
+				<value>hidden</value>
+				<type>String</type>
+			</property>
```
Accidental indentation change

acrowthe committed:
CARDS-1983: Add related form link
Address code style concerns

acrowthe committed:
CARDS-1983: Add related form link
- Improve related form display

sashaandjic commented:
Question #2: Can we add property `Label` to formReference node automatically.  Link with `Related form` text looks more user friendly then `PATIENT XX/ PAUSE-RESUME-STATUS`

acrowthe commented:
Commit pushed since last questions though the timestamp was wrong so conversation is out of order.

Improved related form appearance:
- Prefixed with `Related:`
- Added custom labels for pause and resume forms:
![image](https://user-images.githubusercontent.com/20056751/235794218-cc41b349-0e66-4101-bde9-7d53968f5d0a.png)

- Hid matching subjects from breadscrumbs (custom label disabled for demonstration):
![image](https://user-images.githubusercontent.com/20056751/235794048-37ea08c6-046b-4f44-95b6-7b3ce674c530.png)

Todo:
- Improve recurrent deletion handling
- Add testing instructions for breadcrumbs without shared parents

acrowthe committed:
CARDS-1983: Add related form link
- Improve recursive deletion handling

acrowthe commented:
Pause forms now delete the paired resume form:
![image](https://github.com/data-team-uhn/cards/assets/20056751/62485383-e16f-4835-bc47-f952736f76df)

Resume forms only delete the form reference off of their pause form:
![image](https://github.com/data-team-uhn/cards/assets/20056751/b7ed2bcd-a4c4-45a9-97fd-8967898ac628)

acrowthe commented:
The pop up isn't for the pause form: It's for the formReference node on the pause form since we need to delete the related form link. I added the wording "is referenced in 1 form reference" to try and differentiate that but can change the warning message if you can think of something clearer

sashaandjic commented:
Users have no clue about the ` formReference node`. To them it is just one form that they see on the screen. I would just skip displaying the second dialog all together for at least resumed ones, as user already confirmed that they want to proceed with the deletion. 

Ideally, if possible and not too time consuming, we should say somethin like this:
- for `resumed` forms:  "Do you want to delete this resumed form?" 
- for paused with no resumed pair: "Do you want to delete this paused form?"
- for paused form that has its paired resumed form: "Do you want to delete this paused form and its resumed pair".

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -343,17 +344,35 @@ private void iterateReferrers(
      * @return a string in the format "2 forms, 1 subject(subjectName)" for all traversed nodes
      */
-    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:ExecutableStatementCount", "checkstyle:JavaNCSS"})
```
:(

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -368,6 +387,15 @@ private String listReferrersFromTraversal()
                         break;
+                    case "cards:FormReference":
+                        // If the form reference leads to the parent form being deleted,
+                        // don't list the form reference as a deleted item as it's parent form
```
```suggestion
                        // don't list the form reference as a deleted item as its parent form
```

marta- replied:
```suggestion
                        // don't list the form reference as a deleted item as its parent form
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -394,6 +422,9 @@ private String listReferrersFromTraversal()
                 addNodesToResult(results, "answer", answerCount);
             }
+            if (formReferenceCount > 0) {
+                addNodesToResult(results, "form reference", formReferenceCount);
```
```suggestion
                addNodesToResult(results, "related form", formReferenceCount);
```

marta- replied:
I misunderstood this label. @acrowthe explained and I withdraw the suggestion.

marta- commented:
> Users have no clue about the ` formReference node`. To them it is just one form that they see on the screen. I would just skip displaying the second dialog all together for at least resumed ones, as user already confirmed that they want to proceed with the deletion.

Please correct me id I'm wrong in the following @acrowthe and @sashaandjic.

It was my understanding that the deletion of a Pause form would trigger the deletion of the corresponding Resume form, but the deletion of a resume form would NOT trigger deletion of the corresponding Pause form. When we delete a Resume form, we don't need to mention the reference, or show the second dialog, because the reference would not be deleted.

 
> Ideally, if possible and not too time consuming, we should say something like this:
> 
> * for `resumed` form:  "Do you want to delete this resumed form?"
> * for `paused` with no resumed pair: "Do you want to delete this paused form?"
> * for `paused` form that has its paired `resumed` form: "Do you want to delete this paused form and its resumed pair".
>   For all dialogs we can display the PatienID in the title already, so no need to repeat this info in actual question.

Improving the delete dialog would be great, but these changes would impact other delete buttons. It should be done as a separate task/in a separate PR.

acrowthe committed:
CARDS-1983: Add related form link
Improve UX around recursive deletion when a formReference node
  is the only node being recursively deleted.
Now formReference nodes will not trigger a user prompt to trigger recursive
  deletion or be mentioned if another node triggers a recursive deletion prompt

acrowthe commented:
Deletion improved
Deleting the resume form will skip the recursive deletion prompt.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -364,15 +371,15 @@ private void iterateParentIfRequired(
+     * @return a string in the format "2 forms, 1 subject(subjectName)" for all traversed nodes,
+     *         an empty string if there are no nodes needing confirmation or {@code null} in case of error.
      */
-    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:ExecutableStatementCount", "checkstyle:JavaNCSS"})
```
:)

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -377,6 +403,11 @@ private String listReferrersFromTraversal()
                         break;
+                    case "cards:FormReference":
+                        // Do not add formReferences to the user's prompt:
+                        // The user be prompted to delete the formReferences parent form if it is being deleted
```
```suggestion
                        // The user will be prompted to delete the formReferences parent form if it is being deleted
```

marta- commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/serialize/RelatedFormsProcessor.java:
```diff
@@ -0,0 +1,74 @@
+import io.uhndata.cards.serialize.spi.ResourceJsonProcessor;
+
+/**
+ * Prevent all FormReferences from being exported. The name of this processor is
+ * {@code relatedForms}.
```
The name `relatedForms` is misleading. Without looking at the code or java doc, one could easily assume that the processor somehow includes related forms. If what the processor does is prevent all FormReferences from being exported, I suggest naming it something along the lines of `excludeFormReference`. Other suggestions that make the functionality clear are welcome.

sashaandjic commented:
tested deletion. all good!

sashaandjic reviewed (approved):
tested

acrowthe committed:
CARDS-1983: Add related form link
Major refactor of form references code:
- Add support for multiple form references on a given form
- Move form reference properties into a seperate node
- Add generic way for a form reference to specify that the target form should
reference this form back rather than hardcoding that behaviour into pause forms

acrowthe committed:
CARDS-1983: Add related form link
Address comments

marta- commented:
@sashaandjic please retest

sashaandjic reviewed (approved):
re-tested

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,200 @@
+
+        try {
+            Node formReferences;
+            if (formToModify.hasNode("formReferencess")) {
```
```suggestion
            if (formToModify.hasNode("formReferences")) {
```

IntegralProgrammer replied:
To prevent this kind of bug from happening, you define string constants:

 - `FORM_REFERENCES_NAME = "formReferences"`
 - `FORM_REFERENCES_CARDS_TYPE = "cards:FormReferences"`

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    private void addPropertiesFromProperties(Node newProperties, Node existingProperties)
```
I think it would be better if this function were renamed to `copyNodeLinkbackProperties` since it's only the `LINKBACK_PROPERTIES` that get copied and they get copied from one node to another.

IntegralProgrammer commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -834,3 +837,52 @@
+  // If false, only delete the formReference
+  - recursiveDeleteParent (BOOLEAN) = false mandatory autocreated
+
+  // TODO: should the node version the parent form when modified
+  // See OPV property here:
+  // https://developer.adobe.com/experience-manager/reference-materials/spec/jcr/2.0/25_Appendix.html
```
What does this mean? Does anything need to be implemented?

acrowthe replied:
Future improvement that can be made: A way to flag form reference changes from being versioning exempt meaning they won't need to be checked in and out.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -343,12 +351,36 @@ private void iterateReferrers(
     }
 
+    /**
+     * Recursively call a function all children of the current node's parent if required.
```
```suggestion
     * Recursively call a function on all children of the current node's parent if required.
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,200 @@
+            return;
+        }
+
+        Node formToReference = existingFormReference.getParent().getParent();
```
The purpose of `.getParent().getParent()` is a little bit difficult to understand. Would it be possible to use the `public Node getForm(final Node answer)` method from https://github.com/data-team-uhn/cards/blob/0867e77e7e5b75d388129e475f1e379b8ea04ae5/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormUtilsImpl.java#L96-L108 instead?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -343,12 +351,36 @@ private void iterateReferrers(
+     * @param consumer the function to be called on each node
+     * @throws RepositoryException if any function call fails due to repository errors
+     */
+    private void iterateParentIfRequired(
```
I think this should be renamed as the name somewhat implies that this function returns an _Iterator_ even though this function has a `void` return type and thus doesn't return anything.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/DeleteServlet.java:
```diff
@@ -343,12 +351,36 @@ private void iterateReferrers(
+                // Delete the parent form if specified by the form reference properties
+                iterateChildren(node.getParent().getParent(), consumer, true);
+            } else if (node.getParent().getNodes().getSize() == 1) {
+                // Delete the set of form references if this is the only form reference
```
From what I understand, this `iterateParentIfRequired` function applies an arbitrary function (`consumer`) to all nodes that are iterated through, yet these comments imply that the function will be one that deletes _Form References_.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,200 @@
+        return false;
+    }
+
+    private void checkin(final Node form) throws RepositoryException
```
Shouldn't this be renamed to `checkinIfNeeded` as it performs a check on the checked-in / checked-out status of the Form just like `checkoutIfNeeded` does?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,200 @@
+        if (!(path.contains("formReferences/")
+            && path.substring(0, path.lastIndexOf("/")).endsWith("formReferences"))) {
+            return;
+        }
```
Wouldn't it be easier and more reliable just to check if the `jcr:primaryType` is `cards:FormReference`?

acrowthe committed:
CARDS-1983: Add related form link
Address review comments
- Improve function naming and general code readability

sashaandjic commented:
re-tested - all good
![image](https://github.com/data-team-uhn/cards/assets/23243244/a0270c41-bf81-4201-af67-d109187dbe5e)

IntegralProgrammer commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/PauseResumeFormEditor.java:
```diff
@@ -238,18 +241,63 @@ private void createAnswer(final String questionUUID, final String value)
+                "cards/ReferenceProperties");
+            linkbackProperties.setProperty("label", "Resume Form", Type.STRING);
+
+            setTypeProperties(reference, "cards:FormReference", "cards/Resource", "cards/FormReference");
```
Is there a reason why this is called down here as opposed to immediately after `setDefaultProperties(reference)`?

IntegralProgrammer commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/PauseResumeFormEditor.java:
```diff
@@ -238,18 +241,63 @@ private void createAnswer(final String questionUUID, final String value)
+        }
+    }
+
+    private void setDefaultProperties(NodeBuilder node)
+    {
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
+        node.setProperty(CREATED_PROP, dateFormat.format(new Date()), Type.DATE);
+        node.setProperty("jcr:createdBy", this.rrp.getThreadResourceResolver().getUserID(), Type.NAME);
+    }
+
+    private void setTypeProperties(NodeBuilder node, String primaryType, String superType, String type)
+    {
+        node.setProperty("jcr:primaryType", primaryType, Type.NAME);
+        node.setProperty("sling:resourceSuperType", superType, Type.STRING);
+        node.setProperty("sling:resourceType", type, Type.STRING);
+    }
```
These methods are good, but I think it would make for nicer code if you created some other methods which would, in turn, call these methods. So I recommend creating methods such as `NodeBuilder addReferencePropertiesNode(NodeBuilder node)` which would call `node.setChildNode("referenceProperties")`, then `setDefaultProperties`, and finally `setTypeProperties` on the `NodeBuilder` object that was created with the call to `node.setChildNode("referenceProperties")`.

IntegralProgrammer commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/PauseResumeFormEditor.java:
```diff
@@ -238,18 +241,63 @@ private void createAnswer(final String questionUUID, final String value)
+    {
+        try {
+            NodeBuilder references;
+            if (this.currentNodeBuilder.hasChildNode("formReferences")) {
+                references = this.currentNodeBuilder.getChildNode("formReferences");
+            } else {
+                references = this.currentNodeBuilder.setChildNode("formReferences");
+                setDefaultProperties(references);
+                setTypeProperties(references, "cards:FormReferences", "cards/Resource", "cards/FormReferences");
+            }
```
I think the code organization and readability would be much improved if this code were moved into a `getOrCreateFormReferencesNodeBuilder` method.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,201 @@
+
+        try {
+            Node formReferences;
+            if (formToModify.hasNode(FORM_REFERENCES_NAME)) {
+                formReferences = formToModify.getNode(FORM_REFERENCES_NAME);
+            } else {
+                formReferences = formToModify.addNode(FORM_REFERENCES_NAME);
+                formReferences.setPrimaryType(FORM_REFERENCES_CARDS_TYPE);
+            }
```
I recommend moving this code into a new `Node getOrCreateFormReferencesNode(Node form)` method.

acrowthe committed:
CARDS-1983: Add related form link
Improve code readability

sashaandjic reviewed (approved):
tested

acrowthe committed:
CARDS-2325: Add OPV to related form links

IntegralProgrammer commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/serialize/ExcludeFormReferencesProcessor.java:
I wonder if this would be best if moved to outside of the `heracles-resources` directory as the `cards:FormReferences` data type is defined outside of `heracles-resources` and therefore could potentially be used, in the future, for projects that are not HERACLES. In such cases I'd think it would be likely that we would want to exclude _Form References_ in data exports. @sdumitriu?

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -467,6 +485,33 @@ function Form (props) {
+          {Object.entries(data['formReferences']).filter(([key, value]) => "cards:FormReference" == value['jcr:primaryType']).length > 1 ?
+            <>
+              <br/>
+              {Object.entries(data['formReferences']).filter(([key, value]) => "cards:FormReference" == value['jcr:primaryType']).map(([key, value]) => formReferenceDisplay(value[1]))}
+            </>
+            :
+            formReferenceDisplay(Object.entries(data['formReferences']).filter(([key, value]) => "cards:FormReference" == value['jcr:primaryType'])[0][1])
```
Small code readability suggestion - instead of using hard-coded indices of `[0]` and `[1]`, define constants for `0` and `1` with names that describe what they mean.

acrowthe committed:
CARDS-1983: Add related form link
- Move exclude form reference processor from heracles to base cards
- Code readability adjustment

IntegralProgrammer commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -347,6 +349,24 @@ function Form (props) {
+        break;
+      }
+    }
+    return distinctHierarchy;
```
Calling the `.concat` method on an array doesn't change the array but rather returns a _new_ array with the new elements appended (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat). 

Therefore this method will always return `[]`.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/FormReferenceListener.java:
```diff
@@ -0,0 +1,207 @@
+            checkinNeeded = checkoutIfNeeded(formToModify);
+            formToModify.getSession().save();
+        }
+        if (checkinNeeded) {
+            checkinIfNeeded(formToModify);
+        }
```
Should this be placed in a `finally` block? If `formToModify.getSession().save()` fails wouldn't we always want to be sure that the form pointed to by `formToModify` is checked back in?

@sdumitriu WDTY?

IntegralProgrammer commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -793,6 +793,9 @@
 
   // Children
 
+  // A link to another form
```
Maybe this should be changed to
```suggestion
  // A The links to other forms
```
since, if I understand correctly, it should be possible to link to more than one other form?

sdumitriu committed:
CARDS-1983: Add support for links between resources
New links module

acrowthe committed:
CARDS-1983: Add related form link
- Fix UI issues related to form references with different parents
- Fix browser console log due to missing key in list

sdumitriu committed:
CARDS-2376: Add paused/resumed links between pairs of Pause/Resume Status forms

sdumitriu committed:
CARDS-1983: Add support for links between resources
Adapt the Delete servlet to work with the new links module

acrowthe commented on heracles-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/LinkDefinitions/pausedIn.xml:
```diff
@@ -0,0 +1,98 @@
+-->
+
+<node>
+  <name>pausedIn</name>
```
Not sure I get the name. Paused In what? `pausedLink` seems more descriptive

sdumitriu replied:
The user-facing label is actually `Pause form`, WDYT about `pauseForm`?

marta- replied:
It's the same naming pattern as "createdOn", although I can see why it looks confusing. I don't like it less than "pausedLink".

marta- replied:
pauseForm is perfect.

acrowthe commented on heracles-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/LinkDefinitions/resumedIn.xml:
```diff
@@ -0,0 +1,93 @@
+-->
+
+<node>
+  <name>resumedIn</name>
```
Same name comment: `resumedLink`

sdumitriu committed:
CARDS-1983: Add support for links between resources
Cleanup no longer used code

sdumitriu committed:
CARDS-1983: Add support for links between resources
Renamed Links to LinkUtils
Fixed "since" version
Javadoc update

sdumitriu committed:
[misc] Minor java refactoring

sdumitriu committed:
CARDS-1983: Add support for links between resources
Bugfix: when recursive deleting a linking resource, make sure it is indeed recursively deleted

sdumitriu committed:
CARDS-1983: Add support for links between resources
Bugfix: strong links cannot be added on a checked in form

sdumitriu committed:
CARDS-1983: Add support for links between resources
Bugfix: two backlinks are created instead of one for forced backlinks

sdumitriu committed:
CARDS-1983: Add support for links between resources
Some javadoc improvements

sdumitriu committed:
CARDS-1983: Add support for links between resources
Implementation changes to actually follow the javadocs

sdumitriu committed:
CARDS-1983: Add support for links between resources
Added toString overrides

sdumitriu committed:
CARDS-1983: Add support for links between resources
Add more details in the JSON representation of links, needed in the UI

sdumitriu committed:
CARDS-1983: Add support for links between resources
Update the frontend to work with the new links implementation

sdumitriu committed:
CARDS-1983: Add support for links between resources
Cleanup no longer needed code

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -488,6 +488,22 @@ function Form (props) {
+          {validLinks.length == 1 ?
+              validLinks.map(link => <Link key={link["@name"]} to={"/content.html" + link["to"]}>{link["resourceLabel"]}</Link>)
+              :
+              <List>
```
The fastest way to reduce padding without redesigning the UI is to make the list `dense` and maybe also `disablePadding`.
```suggestion
              <List dense disablePadding>
```

marta- reviewed (approved):
The code looks fine to me. The current UI works well for Heracles where we have a single form link. If we don't have time now to agree on and to implement a better display for multiple links, we can merge as is and revisit that use case later.

marta- committed:
CARDS-1983: Add related form link

Minor UI improvement - reduce spacing in lists of related links.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -496,7 +496,7 @@ function Form (props) {
               validLinks.map(link => <Link key={link["@name"]} to={"/content.html" + link["to"]}>{link["resourceLabel"]}</Link>)
               :
-              <List>
+              <List dense disablePadding>
```
This doesn't change much, but it's a worthy improvement.

![image](https://github.com/data-team-uhn/cards/assets/88663/84cc25a1-8254-4206-a00b-885da4d8ff07)

acrowthe commented on modules/data-model/links/impl/src/main/java/io/uhndata/cards/links/internal/LinksServlet.java:
```diff
@@ -0,0 +1,111 @@
+import io.uhndata.cards.links.api.LinkUtils;
+
+/**
+ * Servlet that outputs all the Form data for a Questionnaire to a CSV.
```
Outdated javadoc

acrowthe reviewed (commented):
A few minor comment; other than that tested and reviewed

sdumitriu committed:
CARDS-1983: Add support for links between resources
Update since version, copy-pasted javadoc

sdumitriu committed:
CARDS-2376: Add paused/resumed links between pairs of Pause/Resume Status forms
Better names for the pause/resume links
